components/containerd/pod/component.go

```diff
@@ -100,95 +149,170 @@ func (c *component) Metrics(ctx context.Context, since time.Time) ([]components.
-	c.lastPodTime = ts
-	c.lastPods = pods
-	c.lastErr = ""
-	c.lastReason = ""
+	c.lastData = d
+	c.lastMu.Unlock()
 }
 
-type Output struct {
-	Pods []PodSandbox `json:"pods,omitempty"`
+const (
+	defaultSocketFile               = "/run/containerd/containerd.sock"
+	defaultContainerRuntimeEndpoint = "unix:///run/containerd/containerd.sock"
+)
+
+// NOTE
+// DO NOT USE https://github.com/kubernetes/kubernetes/blob/v1.32.0-alpha.0/staging/src/k8s.io/cri-client/pkg/remote_runtime.go yet
+// it fails with
+// "code = Unavailable desc = name resolver error: produced zero addresses"
+
+const (
+	// maxMsgSize use 16MB as the default message size limit.
+	// grpc library default is 4MB
+	maxMsgSize = 1024 * 1024 * 16
+
+	// connection parameters
+	maxBackoffDelay      = 3 * time.Second
+	baseBackoffDelay     = 100 * time.Millisecond
+	minConnectionTimeout = 10 * time.Second
+)
+
+// ref. https://github.com/kubernetes/kubernetes/blob/v1.29.2/pkg/kubelet/cri/remote/remote_runtime.go
+func defaultDialOptions() []grpc.DialOption {
+	cps := grpc.ConnectParams{Backoff: backoff.DefaultConfig}
+	cps.MinConnectTimeout = minConnectionTimeout
+	cps.Backoff.BaseDelay = baseBackoffDelay
+	cps.Backoff.MaxDelay = maxBackoffDelay
+	return []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)),
+		grpc.WithConnectParams(cps),
+		grpc.WithContextDialer(dialUnix),
+		grpc.WithBlock(), //nolint:staticcheck
+	}
 }
 
-func (o *Output) JSON() ([]byte, error) {
-	return json.Marshal(o)
+func dialUnix(ctx context.Context, addr string) (net.Conn, error) {
+	return (&net.Dialer{}).DialContext(ctx, "unix", addr)
 }
 
-func (o *Output) describeReason() string {
-	return fmt.Sprintf("total %d pod sandboxes", len(o.Pods))
+func parseUnixEndpoint(endpoint string) (string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", err
+	}
+	if u.Scheme != "unix" {
+		return "", fmt.Errorf("invalid scheme: %s (only supports 'unix' protocol)", u.Scheme)
+	}
+	return u.Path, nil
 }
 
-func (o *Output) States() ([]components.State, error) {
-	if len(o.Pods) == 0 {
-		return []components.State{{
-			Name:    containerd_pod_id.Name,
-			Health:  components.StateHealthy,
-			Healthy: true,
-			Reason:  "no output",
-		}}, nil
+// createConn creates a new CRI service client.
+// Cannot use "k8s.io/kubernetes/pkg/kubelet/cri/remote.NewRemoteRuntimeService" directly
+// as it causes a bunch of go module errors, importing the whole kubernetes repo.
+// ref. https://github.com/kubernetes-sigs/cri-tools/blob/master/cmd/main.go
+// ref. https://github.com/kubernetes/kubernetes/blob/v1.29.2/pkg/kubelet/cri/remote/remote_runtime.go
+// ref. https://github.com/kubernetes/kubernetes/blob/v1.32.0-alpha.0/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
+func createConn(ctx context.Context, endpoint string) (runtimeapi.RuntimeServiceClient, runtimeapi.ImageServiceClient, *grpc.ClientConn, error) {
```

Comment:
should return conn and error... init runtime service client and image service client by need on caller side.